File: vane-core/src/main/java/org/oddlama/vane/core/config/recipes/SmithingRecipeDefinition.java
package org.oddlama.vane.core.config.recipes;

import java.util.HashMap;
import java.util.Map;
import org.bukkit.Material;
import org.bukkit.NamespacedKey;
import org.bukkit.Tag;
import org.bukkit.inventory.Recipe;
import org.bukkit.inventory.RecipeChoice;
import org.bukkit.inventory.SmithingTransformRecipe;
import org.oddlama.vane.util.ItemUtil;

public class SmithingRecipeDefinition extends RecipeDefinition {

    private String base = null;
    private String addition = null;
    private boolean copyNbt = false;
    private String result = null;

    public SmithingRecipeDefinition(String name) {
        super(name);
    }

    public SmithingRecipeDefinition base(String base) {
        this.base = base;
        return this;
    }

    public SmithingRecipeDefinition base(final Tag<?> tag) {
        return base("#" + tag.key());
    }

    public SmithingRecipeDefinition base(Material material) {
        return base(material.key().toString());
    }

    public SmithingRecipeDefinition addition(String addition) {
        this.addition = addition;
        return this;
    }

    public SmithingRecipeDefinition copyNbt(boolean copyNbt) {
        this.copyNbt = copyNbt;
        return this;
    }

    public SmithingRecipeDefinition addition(final Tag<?> tag) {
        return addition("#" + tag.key());
    }

    public SmithingRecipeDefinition addition(Material material) {
        return addition(material.key().toString());
    }

    public SmithingRecipeDefinition result(String result) {
        this.result = result;
        return this;
    }

    @Override
    public Object toDict() {
        final HashMap<String, Object> dict = new HashMap<>();
        dict.put("Base", this.base);
        dict.put("Addition", this.addition);
        dict.put("CopyNbt", this.copyNbt);
        dict.put("Result", this.result);
        dict.put("Type", "smithing");
        return dict;
    }

    @Override
    public RecipeDefinition fromDict(Object dict) {
        if (!(dict instanceof Map<?, ?>)) {
            throw new IllegalArgumentException(
                "Invalid smithing recipe dictionary: Argument must be a Map<String, Object>!"
            );
        }
        final var dictMap = (Map<?, ?>) dict;
        final Object baseObj = dictMap.containsKey("Base") ? dictMap.get("Base") : dictMap.get("base");
        if (baseObj instanceof String base) {
            this.base = base;
        } else {
            throw new IllegalArgumentException("Invalid smithing recipe dictionary: base must be a string");
        }

        final Object additionObj = dictMap.containsKey("Addition") ? dictMap.get("Addition") : dictMap.get("addition");
        if (additionObj instanceof String addition) {
            this.addition = addition;
        } else {
            throw new IllegalArgumentException("Invalid smithing recipe dictionary: addition must be a string");
        }

        final Object copyNbtObj = dictMap.containsKey("CopyNbt") ? dictMap.get("CopyNbt") : dictMap.get("copyNbt");
        if (copyNbtObj instanceof Boolean copyNbt) {
            this.copyNbt = copyNbt;
        } else {
            throw new IllegalArgumentException("Invalid smithing recipe dictionary: copyNbt must be a bool");
        }

        final Object resultObj = dictMap.containsKey("Result") ? dictMap.get("Result") : dictMap.get("result");
        if (resultObj instanceof String result) {
            this.result = result;
        } else {
            throw new IllegalArgumentException("Invalid smithing recipe dictionary: result must be a string");
        }

        return this;
    }

    @Override
    public Recipe toRecipe(NamespacedKey baseKey) {
        return new SmithingTransformRecipe(
            key(baseKey),
            ItemUtil.itemstackFromString(this.result).getLeft(),
            new RecipeChoice.MaterialChoice(Material.NETHERITE_UPGRADE_SMITHING_TEMPLATE),
            RecipeDefinition.recipeChoice(base),
            RecipeDefinition.recipeChoice(addition),
                copyNbt
        );
    }
}

File: vane-core/src/main/kotlin/org/oddlama/vane/core/config/recipes/SmithingRecipeDefinition.kt
package org.oddlama.vane.core.config.recipes

import org.bukkit.Material
import org.bukkit.NamespacedKey
import org.bukkit.Tag
import org.bukkit.inventory.Recipe
import org.bukkit.inventory.RecipeChoice.MaterialChoice
import org.bukkit.inventory.SmithingTransformRecipe
import org.oddlama.vane.util.ItemUtil

/**
 * Recipe definition for smithing transform recipes.
 *
 * @param name logical recipe name.
 */
class SmithingRecipeDefinition(name: String?) : RecipeDefinition(name) {
    /** Base ingredient definition. */
    private var base: String? = null

    /** Addition ingredient definition. */
    private var addition: String? = null

    /** Whether smithing should copy NBT from base item to result. */
    private var copyNbt = false

    /** Result item definition string. */
    private var result: String? = null

    /** Sets base ingredient definition. */
    fun base(base: String?): SmithingRecipeDefinition {
        this.base = base
        return this
    }

    /** Sets base ingredient from a Bukkit tag. */
    fun base(tag: Tag<*>): SmithingRecipeDefinition {
        return base("#" + tag.key())
    }

    /** Sets base ingredient from a material key. */
    fun base(material: Material): SmithingRecipeDefinition {
        return base(material.key().toString())
    }

    /** Sets addition ingredient definition. */
    fun addition(addition: String?): SmithingRecipeDefinition {
        this.addition = addition
        return this
    }

    /** Sets whether NBT should be copied from base input. */
    fun copyNbt(copyNbt: Boolean): SmithingRecipeDefinition {
        this.copyNbt = copyNbt
        return this
    }

    /** Sets addition ingredient from a Bukkit tag. */
    fun addition(tag: Tag<*>): SmithingRecipeDefinition {
        return addition("#" + tag.key())
    }

    /** Sets addition ingredient from a material key. */
    fun addition(material: Material): SmithingRecipeDefinition {
        return addition(material.key().toString())
    }

    /** Sets result item definition string. */
    fun result(result: String?): SmithingRecipeDefinition {
        this.result = result
        return this
    }

    /** Serializes this recipe definition to dictionary form. */
    override fun toDict(): Any {
        val dict = mutableMapOf<String?, Any?>()
        dict["Base"] = this.base
        dict["Addition"] = this.addition
        dict["CopyNbt"] = this.copyNbt
        dict["Result"] = this.result
        dict["Type"] = "smithing"
        return dict
    }

    /** Loads this recipe definition from dictionary form. */
    override fun fromDict(dict: Any?): RecipeDefinition {
        require(dict is MutableMap<*, *>) { "Invalid smithing recipe dictionary: Argument must be a Map<String, Object>!" }
        val baseObj = if (dict.containsKey("Base")) dict["Base"] else dict["base"]
        if (baseObj is String) {
            this.base = baseObj
        } else {
            throw IllegalArgumentException("Invalid smithing recipe dictionary: base must be a string")
        }

        val additionObj = if (dict.containsKey("Addition")) dict["Addition"] else dict["addition"]
        if (additionObj is String) {
            this.addition = additionObj
        } else {
            throw IllegalArgumentException("Invalid smithing recipe dictionary: addition must be a string")
        }

        val copyNbtObj = if (dict.containsKey("CopyNbt")) dict["CopyNbt"] else dict["copyNbt"]
        if (copyNbtObj is Boolean) {
            this.copyNbt = copyNbtObj
        } else {
            throw IllegalArgumentException("Invalid smithing recipe dictionary: copyNbt must be a bool")
        }

        val resultObj = if (dict.containsKey("Result")) dict["Result"] else dict["result"]
        if (resultObj is String) {
            this.result = resultObj
        } else {
            throw IllegalArgumentException("Invalid smithing recipe dictionary: result must be a string")
        }

        return this
    }

    /** Converts this definition into a Bukkit [SmithingTransformRecipe]. */
    override fun toRecipe(baseKey: NamespacedKey?): Recipe {
        val bk = baseKey ?: throw IllegalArgumentException("baseKey cannot be null")
        val baseDef = base ?: throw IllegalArgumentException("Invalid smithing recipe: base must be set")
        val additionDef = addition ?: throw IllegalArgumentException("Invalid smithing recipe: addition must be set")
        val resultDef = result ?: throw IllegalArgumentException("Invalid smithing recipe: result must be set")

        return SmithingTransformRecipe(
            key(bk),
            ItemUtil.itemstackFromString(resultDef).getLeft()!!,
            MaterialChoice(Material.NETHERITE_UPGRADE_SMITHING_TEMPLATE),
            recipeChoice(baseDef),
            recipeChoice(additionDef),
            copyNbt
        )
    }
}
